### PR TITLE
Version 0.2.0

### PR DESCRIPTION
**Pull request by JamesxX**
An almost complete refactor of version 0.1.1. This version adds the optional flag (#4), along with the addition of an assortment of types that are commonly used.

To do:
- [x] Finish rewriting the manual so its up to date
- [x] Finish adding tests (because I'm paranoid there's an edge case I've missed)
- [x] Fill in changelog

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,10 +1,19 @@
...
+- `Optional` validation type. If a schema yields a validation error, the error is suppressed and the returned value is 'auto'
+- `Choice` validation type. Tested value must be contained within the listed choices.
 
 ## Removed
 
 ## Changed
-
+- fixed error in documentation for string type (previously read that it worked on numbers).
+- Dictionaries that don't have a member that is present in the schema no longer produce an error outside of strict contexts.
+- **(Potentially breaking)** Dictionaries default to empty dictionaries rather than none.
```
Comment (tingerrr) [inline CHANGELOG.md:14]: The change that motivated this is something I wanted to suggest before, but always forgot to do so. This is _definitely_ a breaking change, as most usages of `z.dictioanry` now need additional parentheses.

File: src/types.typ [hunk 2]
```diff
@@ -1,7 +1,22 @@
...
+#import "assertions.typ": one-of
+#import "types/array.typ": array
+#import "types/dictionary.typ": dictionary
+#import "types/logical.typ": either
+#import "types/string.typ": string, ip, email
+#import "types/tuple.typ": tuple
+
+#let any = base-type.with(name: "any")
+#let boolean = base-type.with(name: "bool", types: (bool,))
+#let color = base-type.with( name: "color", types: (type(rgb(0,0,0)),))
```
Comment (tingerrr) [inline src/types.typ:11]: I think the color type can be named using [`color`](https://typst.app/docs/reference/visualize/color/).

File: src/types.typ [hunk 3]
```diff
@@ -1,7 +1,22 @@
...
+#import "types/array.typ": array
+#import "types/dictionary.typ": dictionary
+#import "types/logical.typ": either
+#import "types/string.typ": string, ip, email
+#import "types/tuple.typ": tuple
+
+#let any = base-type.with(name: "any")
+#let boolean = base-type.with(name: "bool", types: (bool,))
+#let color = base-type.with( name: "color", types: (type(rgb(0,0,0)),)) 
+#let content = base-type.with(name: "content", types: (type([]),str))
```
Comment (tingerrr) [inline src/types.typ:12]: As above, `content` is a type that can be named.

File: src/schemas/pubmatter.typ [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        (
+          type: z.content(optional: true),
+          date: z.date(pre-transform: coerce.date)
+        ),
+        pre-transform: coerce.dictionary((it)=>(date: it))
+      ),
+      pre-transform: coerce.array
+    ),
+  )
+)
```
Comment (tingerrr) [inline src/schemas/pubmatter.typ:142]: This seems very specific to publication, which may not be useful to downstream users at all, perhaps this is better left as a supplemental package?

**Review by tingerrr — COMMENTED** [review]
Overall pretty good.

I think it's very important that we add a formatter to the workflow. Some of the stylistic choices make it very hard to review the code. This also makes it harder for others to contribute. [typstyle](https://github.com/Enter-tainer/typstyle) seems to be a good choice.

**Comment by JamesxX** [comment]
> Overall pretty good.
> 
> I think it's very important that we add a formatter to the workflow. Some of the stylistic choices make it very hard to review the code. This also makes it harder for others to contribute. [typstyle](https://github.com/Enter-tainer/typstyle) seems to be a good choice.

Thanks, I've applied typstyle to the current files, but do you know how to install it locally so I can make a just file command?

File: src/types.typ (hunk 2 above)
Comment (JamesxX) [reply src/types.typ]: done #[6f79ed2](https://github.com/JamesxX/valkyrie/pull/5/commits/6f79ed25f8c7c128a1ff3fa3054f920b463a194a)

File: src/types.typ (hunk 3 above)
Comment (JamesxX) [reply src/types.typ]: done #[6f79ed2](https://github.com/JamesxX/valkyrie/pull/5/commits/6f79ed25f8c7c128a1ff3fa3054f920b463a194a)

File: src/schemas/pubmatter.typ (hunk 4 above)
Comment (JamesxX) [reply src/schemas/pubmatter.typ]: Initially, this was going to be upstreamed into pubmatter's package, but the reasoning slowly morphed into the idea of providing well-thought out commonly-used schemas (such as authors, or abstracts) for other templates to implement to get a ball rolling on a form of opt-in standard for template arguments.

It's a good point that it could be in a separate package, but I'd worry about potential version conflicts in the far future. Keeping it in the same package ensures that community schemas are always compatible with the version of valkyrie they ship with.

File: CHANGELOG.md (hunk 1 above)
Comment (JamesxX) [reply CHANGELOG.md:14]: Thanks. I'm considering splitting the changelog into a changelog and a migration guide (or having a migration guide as part of the changelog, I don't know yet).

File: src/schemas/pubmatter.typ (hunk 4 above)
Comment (tingerrr) [reply src/schemas/pubmatter.typ]: > but I'd worry about potential version conflicts in the far future

These will hardly be a problem now. Currently, packages are all pinned and in the far future if we do actually get to have a semver dependency resolver in Typst, it will likely have custom types by then.

I don't think it makes much sense to provide more than the building blocks and actually common types in this package itself.

**Comment by tingerrr** [comment]
> Thanks, I've applied typstyle to the current files, but do you know how to install it locally so I can make a just file command?

If you have a rust toolchain installed, it would be `cargo install --git LINK`, otherwise the releases should have precompiled binaries for your platform of choice. The tinymist LSP also comes with this formatter built in.

File: CHANGELOG.md (hunk 1 above)
Comment (tingerrr) [reply CHANGELOG.md:14]: Migration guide or not, it should definitely be marked as a breaking change, not a potentially breaking change.

Comment (JamesxX) [reply CHANGELOG.md]: done, [d5e26ae](https://github.com/JamesxX/valkyrie/pull/5/commits/d5e26ae602eadfa3e73ff13f5c9a86b203689151)

File: src/schemas/pubmatter.typ (hunk 4 above)
Comment (JamesxX) [reply src/schemas/pubmatter.typ]: How about keeping author and abstracts, and upstreaming affiliations and pubmatter?

Comment (JamesxX) [reply src/schemas/pubmatter.typ]: I've elected to retain only the author schema.

**Comment by JamesxX** [comment]
blocked by https://github.com/typst/packages/pull/694

**Review by tingerrr — CHANGES_REQUESTED** [review]
I think the manual is missing a monospace font, the examples are all non-monospace.

**Comment by JamesxX** [comment]
> I think the manual is missing a monospace font, the examples are all non-monospace. You're also sill referencing a local fork of mantys, we don't wanna have that blocking CI, so we either need to install the fork in CI or switch to a live version.

Looks to be mantys trying to use the font "Liberation Mono" for code for some reason. Not sure how to install it in the CI but the live version doesn't export `add-type`, which is needed to show the proper return types of the functions.

Will recompile docs with font and push that to both this repo and the packages repo.
